### DISCUSSION TITLE
Add missing negative charges to atoms in molecules?

### DISCUSSION THREAD

**Discussion by najwalb** [rdkit/rdkit#6344] [Q&A] [answered]
I am creating molecules as editable objects from graphs containing only atom and bond types. I am trying to add the formal charges, if needed, posthoc. I know one way to add a positive formal charge (to nitrogen for example) is to check if the number of bonds it has exceeds its valence. 

Is there a similar rule I can use to add a negative formal charge?  What's even more confusing in this case is that RDKit does not seem to pick on missing negative charges, unlike missing positive charges. Why is this the case?

Here is an example to illustrate:

```
 # input smiles
smi = 'NC1=CC(Cl)=CC=C1[N+](=O)[O-]'

# my own modeling of molecules 
graph = get_atom_bond_graph(smi) 

# returns an rwmol object from the graph
mol = get_mol_from_graph(graph)

# smi2 is missing all formal charges => cannot be sanitized because the valence of N is off
smi2 = Chem.MolToSmiles(mol) # NC1=CC(Cl)=CC=C1N(=O)O  

# add missing formal charges by checking valence rules 
correct_smiles(mol)

smi3 = Chem.MolToSmiles(mol) # NC1=CC(Cl)=CC=C1[N+](=O)O'

# ... how to get [0-] with posthoc checks?
```

**Comment by greglandrum** [comment]
@najwalb what you're trying to do doesn't really make sense in the context of SMILES: where you need to specify the formal charges on atoms.

For your case, and assuming that we are working with standard organic atoms which need to obey the octet rule, there are three pieces of information you need to completely specify the valence of a particular atom. Assuming you know the type of atom (C, N, O, etc), that info is:
- the formal charge
- the number of radical electrons
- the total valence
Given two pieces of information you can calculate the other one.

Making some simplifications (if you really want to make this work it's more complex...) and assuming that you don't have radicals, then you can generally assign a +1 for each extra bond coming into an atom - so four-coordinate N is +1 since N wants to form three bonds - or -1 for each missing bond - so single-coordinate O without any Hs on it would be -1.

**Reply by najwalb** [reply]
Yes this makes sense, if I know the implicit hydrogens I am able to infer any missing formal charge using valence rules as you said. The problem is that the negative formal charge conflicts with a missing implicit hydrogen. I couldn't find any work around to this (infer the negative formal charge, in the O example you gave without knowing the implicit hydrogens). So for now I started modeling formal charge explicitly. 

Just waiting for your confirmation/comments on this observation before closing the issue.

**Reply by greglandrum** [reply, marked as answer]
As mentioned: you need either the formal charge or the number of implicit Hs. So explicitly tracking formal charges is a good idea
